core/src/main/scala/org/apache/spark/storage/memory/MemoryStore.scala

```diff
@@ -354,63 +401,30 @@ private[spark] class MemoryStore(
       ser.serializeStream(serializerManager.wrapForCompression(blockId, redirectableStream))
     }
 
-    // Request enough memory to begin unrolling
-    keepUnrolling = reserveUnrollMemoryForThisTask(blockId, initialMemoryThreshold, memoryMode)
-
-    if (!keepUnrolling) {
-      logWarning(s"Failed to reserve initial memory threshold of " +
-        s"${Utils.bytesToString(initialMemoryThreshold)} for computing block $blockId in memory.")
-    } else {
-      unrollMemoryUsedByThisBlock += initialMemoryThreshold
+    def storeValue(value: T): Unit = {
+      serializationStream.writeObject(value)(classTag)
     }
 
-    def reserveAdditionalMemoryIfNecessary(): Unit = {
-      if (bbos.size > unrollMemoryUsedByThisBlock) {
-        val amountToRequest = (bbos.size * memoryGrowthFactor - unrollMemoryUsedByThisBlock).toLong
-        keepUnrolling = reserveUnrollMemoryForThisTask(blockId, amountToRequest, memoryMode)
-        if (keepUnrolling) {
-          unrollMemoryUsedByThisBlock += amountToRequest
-        }
-      }
-    }
-
-    // Unroll this block safely, checking whether we have exceeded our threshold
-    while (values.hasNext && keepUnrolling) {
-      serializationStream.writeObject(values.next())(classTag)
-      elementsUnrolled += 1
-      if (elementsUnrolled % memoryCheckPeriod == 0) {
-        reserveAdditionalMemoryIfNecessary()
+    def estimateSize(precise: Boolean): Long = {
+      if (precise) {
+        serializationStream.flush()
```

Comment:
I don't see anywhere in the previous code call `flush`.

Reply:
Because there are some data cached in the serializationStream, we can't get the precise size if don't call `flush`.  Previous we don't check again after unrolled the block, and it directly call the `serializationStream.close()`. But here we maybe need the `serializationStream` again if we can't get anther unroll memory, so we only should call `flush`.

Reply:
can you send a PR to fix this issue for `putIteratorAsBytes` first? It will make this PR easier to review

Reply:
OK, I'll do it tomorrow.

Reply:
@cloud-fan Sorry for the previous saying, I read the code again. Here seems call `serializationStream .close` is also OK. Because the the iterator is has not value need write, that's meaning the `serializationStream` don't need anymore.